src/passes/transform/cse.cpp

```diff
@@ -0,0 +1,174 @@
+#include "coreir/passes/transform/cse.h"
+
+using namespace std;
+using namespace CoreIR;
+
+// Desired CoreIR iterators/functional methods
+// breadth-first tree iterator on wireables
+// depth-first tree iterator on wireables
+// leaf-iterator on wireables (Only single bits)
+// breadth-first iteration on types
+// depth-first iterator on types
+// leaf iterator on types 
+// 'All' operator with a predicate
+// 'Any' operator with a predicate
+// 'Map' operator
+// 'Filter' operator (both a modify and a return new)
+
+
+//Some possible (untested) implementations of some of the functions
+template <typename Cont, typename Pred>
```

Comment:
are `Cont`,  `Pred`, `Fun`, etc. standard abbreviations? It's totally fine if so and I'm not a domain expert (I can guess `Pred` is predicate and `Fun` is function but just looking at the function signature doesn't do much for me in terms of variables). And you have `Cont cont` but `Pred predicate`...keep it consistent between abbreviations and full names.